File: refactor-attendance-structure.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const BASE_DIR = path.join(__dirname, 'server');
const ATTENDANCE_ROUTES = path.join(BASE_DIR, 'routes', 'attendance');

// Source directories
const SRC_MODELS = path.join(ATTENDANCE_ROUTES, 'models');
const SRC_CONTROLLERS = path.join(ATTENDANCE_ROUTES, 'controllers');
const SRC_SERVICES = path.join(ATTENDANCE_ROUTES, 'services');

// Destination directories
const DEST_MODELS = path.join(BASE_DIR, 'model', 'attendance');
const DEST_CONTROLLERS = path.join(BASE_DIR, 'controllers', 'attendance');
const DEST_SERVICES = path.join(BASE_DIR, 'services', 'attendance');

console.log('🔍 LISTING SOURCE FILES\n');
console.log('Models in', SRC_MODELS);
const models = fs.readdirSync(SRC_MODELS);
console.log('  ', models.join(', '));

console.log('\nControllers in', SRC_CONTROLLERS);
const controllers = fs.readdirSync(SRC_CONTROLLERS);
console.log('  ', controllers.join(', '));

console.log('\nServices in', SRC_SERVICES);
const services = fs.readdirSync(SRC_SERVICES);
console.log('  ', services.join(', '));

console.log('\n📁 CREATING DESTINATION DIRECTORIES\n');

// Create destination directories
[DEST_MODELS, DEST_CONTROLLERS, DEST_SERVICES].forEach(dir => {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
    console.log('✓ Created:', dir);
  } else {
    console.log('✓ Already exists:', dir);
  }
});

console.log('\n📋 COPYING FILES\n');

// Function to copy file
function copyFile(src, dest, type) {
  const content = fs.readFileSync(src, 'utf8');
  fs.writeFileSync(dest, content);
  console.log(`✓ Copied ${type}: ${path.basename(src)}`);
  return dest;
}

// Copy models
const copiedModels = models.map(file => 
  copyFile(path.join(SRC_MODELS, file), path.join(DEST_MODELS, file), 'model')
);

// Copy controllers
const copiedControllers = controllers.map(file =>
  copyFile(path.join(SRC_CONTROLLERS, file), path.join(DEST_CONTROLLERS, file), 'controller')
);

// Copy services
const copiedServices = services.map(file =>
  copyFile(path.join(SRC_SERVICES, file), path.join(DEST_SERVICES, file), 'service')
);

console.log('\n🔧 UPDATING IMPORT PATHS\n');

// Function to update imports in a file
function updateImports(filePath, type) {
  let content = fs.readFileSync(filePath, 'utf8');
  let updated = false;

  // Replace relative model imports
  if (type === 'model' || type === 'controller' || type === 'service') {
    // Match patterns like require('../models/X') or require('./models/X')
    const modelImportRegex = /require\(['"`]\.\.\/models\/([^'"`]+)['"`]\)/g;
    if (modelImportRegex.test(content)) {
      content = content.replace(modelImportRegex, "require('../../model/attendance/$1')");
      updated = true;
    }

    // Special case for User model
    const userImportRegex = /require\(['"`]\.\.\/models\/User['"`]\)/g;
    if (userImportRegex.test(content)) {
      content = content.replace(userImportRegex, "require('../../model/userModel.mjs')");
      updated = true;
    }
  }

  if (updated) {
    fs.writeFileSync(filePath, content);
    console.log(`✓ Updated imports in: ${path.basename(filePath)}`);
  }
}

// Update imports in copied files
copiedModels.forEach(file => updateImports(file, 'model'));
copiedControllers.forEach(file => updateImports(file, 'controller'));
copiedServices.forEach(file => updateImports(file, 'service'));

console.log('\n🔄 UPDATING ROUTE FILES\n');

// Update route files to import controllers from new location
const routeFiles = [
  path.join(ATTENDANCE_ROUTES, 'attendanceRoutes.mjs'),
  path.join(ATTENDANCE_ROUTES, 'hodRoutes.mjs'),
  path.join(ATTENDANCE_ROUTES, 'leaveRoutes.mjs'),
  path.join(ATTENDANCE_ROUTES, 'masterRoutes.mjs')
];

routeFiles.forEach(routeFile => {
  if (fs.existsSync(routeFile)) {
    let content = fs.readFileSync(routeFile, 'utf8');
    let updated = false;

    // Update controller imports: change ./controllers/X to ../../../controllers/attendance/X
    const controllerRegex = /require\(['"`]\.\/controllers\/([^'"`]+)['"`]\)/g;
    if (controllerRegex.test(content)) {
      content = content.replace(controllerRegex, "require('../../../controllers/attendance/$1')");
      updated = true;
    }

    if (updated) {
      fs.writeFileSync(routeFile, content);
      console.log(`✓ Updated: ${path.basename(routeFile)}`);
    } else {
      console.log('○ No changes needed in: ' + path.basename(routeFile));
    }
  } else {
    console.log('⚠ File not found:', routeFile);
  }
});

console.log('\n🗑️  CLEANING UP OLD DIRECTORIES\n');

// Remove old directories
[SRC_MODELS, SRC_CONTROLLERS, SRC_SERVICES].forEach(dir => {
  try {
    // Remove files first
    if (fs.existsSync(dir)) {
      const files = fs.readdirSync(dir);
      files.forEach(file => {
        fs.unlinkSync(path.join(dir, file));
      });
      fs.rmdirSync(dir);
      console.log('✓ Deleted:', dir);
    }
  } catch (err) {
    console.log('⚠ Error deleting', dir, ':', err.message);
  }
});

console.log('\n✅ REFACTORING COMPLETE!\n');
console.log('Summary:');
console.log(`  - Copied ${models.length} model files`);
console.log(`  - Copied ${controllers.length} controller files`);
console.log(`  - Copied ${services.length} service files`);
console.log(`  - Updated imports in all copied files`);
console.log(`  - Updated ${routeFiles.filter(f => fs.existsSync(f)).length} route files`);
console.log(`  - Deleted old attendance subdirectories`);

File: refactor_attendance_structure.py
import os
import shutil
from pathlib import Path

BASE_DIR = Path("C:\\Users\\india\\Desktop\\Projects\\eximdev\\server")

# Source directories
SRC_MODELS = BASE_DIR / "routes" / "attendance" / "models"
SRC_CONTROLLERS = BASE_DIR / "routes" / "attendance" / "controllers"
SRC_SERVICES = BASE_DIR / "routes" / "attendance" / "services"

# Destination directories
DEST_MODELS = BASE_DIR / "model" / "attendance"
DEST_CONTROLLERS = BASE_DIR / "controllers" / "attendance"
DEST_SERVICES = BASE_DIR / "services" / "attendance"

print("🔍 LISTING SOURCE FILES\n")

print(f"Models in {SRC_MODELS}")
models = list(SRC_MODELS.glob("*"))
model_names = [m.name for m in models if m.is_file()]
print("  ", ", ".join(model_names))

print(f"\nControllers in {SRC_CONTROLLERS}")
controllers = list(SRC_CONTROLLERS.glob("*"))
controller_names = [c.name for c in controllers if c.is_file()]
print("  ", ", ".join(controller_names))

print(f"\nServices in {SRC_SERVICES}")
services = list(SRC_SERVICES.glob("*"))
service_names = [s.name for s in services if s.is_file()]
print("  ", ", ".join(service_names))

print("\n📁 CREATING DESTINATION DIRECTORIES\n")

# Create destination directories
for dest_dir in [DEST_MODELS, DEST_CONTROLLERS, DEST_SERVICES]:
    if not dest_dir.exists():
        dest_dir.mkdir(parents=True, exist_ok=True)
        print(f"✓ Created: {dest_dir}")
    else:
        print(f"✓ Already exists: {dest_dir}")

print("\n📋 COPYING FILES\n")

copied_models = []
for model_file in models:
    if model_file.is_file():
        dest_file = DEST_MODELS / model_file.name
        shutil.copy2(model_file, dest_file)
        copied_models.append(dest_file)
        print(f"✓ Copied model: {model_file.name}")

copied_controllers = []
for controller_file in controllers:
    if controller_file.is_file():
        dest_file = DEST_CONTROLLERS / controller_file.name
        shutil.copy2(controller_file, dest_file)
        copied_controllers.append(dest_file)
        print(f"✓ Copied controller: {controller_file.name}")

copied_services = []
for service_file in services:
    if service_file.is_file():
        dest_file = DEST_SERVICES / service_file.name
        shutil.copy2(service_file, dest_file)
        copied_services.append(dest_file)
        print(f"✓ Copied service: {service_file.name}")

print("\n🔧 UPDATING IMPORT PATHS\n")

def update_imports(file_path):
    """Update relative imports in copied files"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        original_content = content
        
        # Replace model imports: ../models/X -> ../../model/attendance/X
        import re
        
        # Pattern 1: require('../models/X') or similar
        content = re.sub(
            r"require\(['\"`]\.\.\/models\/([^'\""`]+)['\""`]\)",
            r"require('../../model/attendance/\1')",
            content
        )
        
        # Pattern 2: Special case for User model
        content = re.sub(
            r"require\(['\"`]\.\.\/models\/User['\""`]\)",
            r"require('../../model/userModel.mjs')",
            content
        )
        
        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            print(f"✓ Updated imports in: {file_path.name}")
    except Exception as e:
        print(f"⚠ Error updating {file_path.name}: {e}")

# Update imports in copied files
for f in copied_models + copied_controllers + copied_services:
    update_imports(f)

print("\n🔄 UPDATING ROUTE FILES\n")

route_files = [
    BASE_DIR / "routes" / "attendance" / "attendanceRoutes.mjs",
    BASE_DIR / "routes" / "attendance" / "hodRoutes.mjs",
    BASE_DIR / "routes" / "attendance" / "leaveRoutes.mjs",
    BASE_DIR / "routes" / "attendance" / "masterRoutes.mjs"
]

for route_file in route_files:
    if route_file.exists():
        try:
            with open(route_file, 'r', encoding='utf-8') as f:
                content = f.read()
            
            original_content = content
            
            import re
            # Update controller imports: ./controllers/X -> ../../../controllers/attendance/X
            content = re.sub(
                r"require\(['\"`]\.\/controllers\/([^'\""`]+)['\""`]\)",
                r"require('../../../controllers/attendance/\1')",
                content
            )
            
            if content != original_content:
                with open(route_file, 'w', encoding='utf-8') as f:
                    f.write(content)
                print(f"✓ Updated: {route_file.name}")
            else:
                print(f"○ No changes needed in: {route_file.name}")
        except Exception as e:
            print(f"⚠ Error updating {route_file.name}: {e}")
    else:
        print(f"⚠ File not found: {route_file}")

print("\n🗑️  CLEANING UP OLD DIRECTORIES\n")

# Remove old directories
for old_dir in [SRC_MODELS, SRC_CONTROLLERS, SRC_SERVICES]:
    try:
        if old_dir.exists():
            shutil.rmtree(old_dir)
            print(f"✓ Deleted: {old_dir}")
    except Exception as e:
        print(f"⚠ Error deleting {old_dir}: {e}")

print("\n✅ REFACTORING COMPLETE!\n")
print("Summary:")
print(f"  - Copied {len(model_names)} model files")
print(f"  - Copied {len(controller_names)} controller files")
print(f"  - Copied {len(service_names)} service files")
print(f"  - Updated imports in all copied files")
print(f"  - Updated route files")
print(f"  - Deleted old attendance subdirectories")
